### PR TITLE
Re-organize tests to avoid shell processes (branch: feature/reorg_tests)

### PR DESCRIPTION
**Pull request by kbourgoin**
I'm opening this test for a review of the approach, not because all the tests have been reorganized. If there's support for the ideas in here, then I'll finish it up and replace our current tests with this.

## Overview

Right now, our tests are a bit brittle, the barrier to entry for adding new tests is too high, and debugging failing tests, or gathering code coverage statistics is fairly difficult. This is by no means to say the current tests are *bad*. They work, they test things, and they aren't *too* hard to add new things to. I did this because I want adding new tests to be **stupid easy**. Everything that makes adding new tests easier to write and debug means that more and better tests will get written.

To do this, the big thing I wanted to do was remove the subprocess calls entirely. While they look like Storm, they aren't Storm, and they add a good bit of complication to the setup. For one, it makes using pdb to debug a test impossible, and the break in process space makes code coverage harder to collect.

Finally, we didn't have true integration tests. True integration tests would integrate with Storm and verify that communication was successful. Instead, it integrated with a mock that we created that mimics Storm. Instead of using a subprocess to mock Storm, I'm opting for the mock library to mock individual components. The mock library is a common Python tool for testing that any contributor walking in will know. Using a common tool like that will remove the barrier or learning how our subprocess storm mock works, which makes writing tests that much easier.

*Disclaimer: I've no idea how this will play with python3. Testing that will come later.*

## Benefits

* Easier debugging into running test processes
* Significantly faster tests
  * No process creation/teardown for every test
* Simpler testing
  * Use common mock tools instead of our own mock framework
* Unit testing means more focused tests covering specific use cases

## Next steps

If other people agree this is a good approach, I'll finish test coverage for the current streamparse version. @msukmanowsky has a large Pull Request coming in, so I'll wait until that's resolved and write all the necessary testing for what's there. Once that's complete, I'll hook up coveralls for streamparse to get code coverage stats. I don't know how useful they'll be, but I'm curious about it and it's integration with Travis-CI.

### CODE REVIEW

**Comment by dan-blanchard** [comment]
@kbourgoin The tests here are intended to replace those in `tests/ipc`, correct?

**Comment by kbourgoin** [comment]
Wow. I had forgotten this even existed.  Yeah, that was the goal.

**Comment by dan-blanchard** [comment]
Okay, cool.  I definitely prefer the mock approach, so I'll put the finishing touches on this.

File: test/streamparse/test_bolt.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+        process_batch_mock.assert_has_calls([
+            mock.call(0, [self.tups[0], self.tups[2]]),
+            mock.call(1, [self.tups[1]]),
+        ], any_order=True)
+
+    def _test_exception_handling(self):
+        # Make sure the exception gets from the worker thread to the main
+        for i in range(1):
+            self.bolt._run()
+        self.assertRaises(NotImplementedError, lambda: time.sleep(0.1))
```
Comment (dan-blanchard) [inline test/streamparse/test_bolt.py]: @kbourgoin Why would this work? I can't find anything in the Python docs saying that `time.sleep()` would ever throw a `NotImplementedError`.

Comment (dfdeshom) [inline test/streamparse/test_bolt.py]: @dan-blanchard note that this method doesn't run when calling `unittest.main()` on the file. If you rename it to `test_exception_handling()` , it will run and fail

Comment (dan-blanchard) [inline test/streamparse/test_bolt.py]: @dfdeshom Thanks for pointing that out! I'm still confused by it though, since later in the file there are a couple lines like:

```python
try: time.sleep(0.1)
except NotImplementedError: pass
```

which again imply that somehow `time.sleep()` is going to raise a `NotImplementedError`, which is strange.

Comment (kbourgoin) [inline test/streamparse/test_bolt.py]: It's been a while since I wrote this, but the except is there because the exception is thrown in a background thread.  These look like tests for the BatchingBolt, which has a background thread to read and write tuples.

I'm not 100% sure what this thread did, or why I'm even testing for NotImplementedError.  At a guess, it was for meaningless test coverage.  Other tests with this could be more legit, however.

**Comment by dan-blanchard** [comment]
@kbourgoin Since I sort of stole this from you, could you give this a quick review?

**Comment by dan-blanchard** [comment]
Hmm... making the sleeps longer didn't fix the sporadic crashing with BatchingBolt tests, so I'm going to try to get mocking _handle_worker_exception working.

**Comment by kbourgoin** [comment]
Nice work with the coveralls, btw.  I want to hook that up for PyKafka as well.

**Comment by kbourgoin** [comment]
@dan-blanchard i'll take a look Monday. My brain is too far away from this to do a review today.

File: .travis.yml [hunk 2]
```diff
@@ -7,9 +7,16 @@ python:
...
+  - python setup.py --version
 
-script: nosetests
+# Run test
+script:
+  - nosetests -v --with-cov --cov streamparse --cov-config .coveragerc --logging-level=DEBUG
+
+# Calculate coverage on success
+after_success:
+  - coveralls --config_file .coveragerc
```
Comment (dan-blanchard) [inline .travis.yml]: I've been deleting them, but we're actually getting duplicate coveralls comments because it's commenting for every single Travis job that succeeds (i.e., 2.7, 3.3, 3.4, and pypy), so we should probably just pick a version and have coveralls only get called for that version of Python.

File: streamparse/__init__.py [hunk 3]
```diff
@@ -6,14 +6,12 @@
...
-import streamparse.base
-import streamparse.bolt
+import streamparse.storm
 import streamparse.cmdln
 import streamparse.contextmanagers
 import streamparse.debug
 import streamparse.decorators
-import streamparse.ipc
-import streamparse.spout
+from streamparse.storm import bolt, component, spout
```
Comment (dan-blanchard) [inline streamparse/__init__.py]: I did this to try to not break the API for most users (although the rare person directly referencing `streamparse.ipc` will be out of luck).

File: streamparse/storm/component.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+            raise
+
+    def flush(self):
+        """No-op method to prevent crashes when someone does
+        sys.stdout.flush.
+        """
+        pass
+
+
+Tuple = namedtuple('Tuple', 'id component stream task values')
```
Comment (dan-blanchard) [inline streamparse/storm/component.py]: When we were explicitly defining `Tuple` as just a class that inherited from `object` (and set `__slots__`), we didn't set things up such that `__eq__` worked, so this seemed a simpler approach to getting what we wanted (since `namedtuple` classes always have `__slots__` set).

File: test/streamparse/test_bolt.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+                         'comp': 'some_spout',
+                         'stream': 'default',
+                         'task': 'some_bolt',
+                         'tuple': [1, 2, 3]}
+        tup_json = "{}\nend\n".format(json.dumps(self.tup_dict))
+        self.tup = Tuple(self.tup_dict['id'], self.tup_dict['comp'],
+                         self.tup_dict['stream'], self.tup_dict['task'],
+                         self.tup_dict['tuple'],)
+        self.bolt = Bolt(input_stream=itertools.cycle(tup_json.splitlines(True)),
+                         output_stream=BytesIO())
```
Comment (dan-blanchard) [inline test/streamparse/test_bolt.py]: I don't use the `output_stream` bit in any tests yet, but if we wanted to, we could double-check that the write things end up in stdout by checking `self.bolt.output_stream.getvalue()`.

File: .travis.yml (hunk 2 above)
Comment (kbourgoin) [inline .travis.yml]: My vote is either 2.7 or 3.4.  If we want to live in the future, we can just do 3.4.  The coverage information ought to be the same for them all anyway, no?

Comment (dan-blanchard) [inline .travis.yml]: 3.4 sounds good to me.  The only reason they would be different is if we had different code paths based on the Python version, but we don't have much of that.  For SKLL, the 2.7 and and 3.x versions due do a couple things rather differently, so we record coverage for both, duplicate comments be damned.

File: streamparse/storm/bolt.py [hunk 6]
```diff
@@ -300,10 +304,10 @@ def process_batch(self, key, tups):
     Note: Can be fractional to specify greater precision (e.g. 2.5).
     """
 
-    def __init__(self):
-        super(BatchingBolt, self).__init__()
+    def __init__(self, *args, **kwargs):
+        super(BatchingBolt, self).__init__(*args, **kwargs)
         self.exc_info = None
-        signal.signal(signal.SIGINT, self._handle_worker_exception)
+        signal.signal(signal.SIGUSR1, self._handle_worker_exception)
```
Comment (kbourgoin) [inline streamparse/storm/bolt.py]: Why the change from SIGINT to SIGUSR1?

File: streamparse/storm/component.py (hunk 4 above)
Comment (kbourgoin) [inline streamparse/storm/component.py]: :+1:

File: streamparse/storm/bolt.py (hunk 6 above)
Comment (dan-blanchard) [inline streamparse/storm/bolt.py]: I was running tests and got irritated that CTRL-C would cause an exception in `_handle_exception`, because `self.exc_info` wouldn't be set.  It made me realize that what we really wanted was a signal that didn't conflict with anything else.

**Comment by kbourgoin** [comment]
It has been a *very* long time since I started down this path, but this all looks good to me.  Although, for a testing rewrite branch, we should probably get Travis green before we merge. :)

File: streamparse/storm/bolt.py (hunk 6 above)
Comment (kbourgoin) [inline streamparse/storm/bolt.py]: Yeah, that makes a lot of sense.  I'm cool with that.  There's still the fact that a BatchingBolt has to exist on the main thread, or signal handlers don't attach properly, but I guess that hasn't been an issue for anyone yet.

**Comment by dan-blanchard** [comment]
> we should probably get Travis green before we merge

It was before I merged in master, which is weird, because the changes there were really small.  I'll figure out why `fail_mock` is getting called an extra time in `test_auto_fail_partial` now.

File: .travis.yml (hunk 2 above)
Comment (dan-blanchard) [inline .travis.yml]: Okay, I just fixed that, and once the build succeeds, I'll merge this.
